File: modules/core-api/src/main/java/com/enonic/xp/content/page/PageTemplate.java
package com.enonic.xp.content.page;


import java.util.ArrayList;

import com.enonic.xp.content.Content;
import com.enonic.xp.content.ContentId;
import com.enonic.xp.data.Property;
import com.enonic.xp.data.PropertyTree;
import com.enonic.xp.data.Value;
import com.enonic.xp.schema.content.ContentTypeName;
import com.enonic.xp.schema.content.ContentTypeNames;

public final class PageTemplate
    extends Content
{
    private final PageTemplateKey key;

    private PageTemplate( final Builder builder )
    {
        super( builder );
        this.key = builder.key;
    }

    public PageTemplateKey getKey()
    {
        return key;
    }

    public PropertyTree getConfig()
    {
        return getPage().getConfig();
    }

    public DescriptorKey getController()
    {
        if ( this.getPage() == null )
        {
            return null;
        }
        return this.getPage().getController();
    }

    public ContentTypeNames getCanRender()
    {
        final ArrayList<ContentTypeName> list = new ArrayList<>( this.getData().countNames( "supports" ) );
        for ( final Property property : this.getData().getProperties( "supports" ) )
        {
            if ( !property.hasNullValue() )
            {
                list.add( ContentTypeName.from( property.getString() ) );
            }
        }
        return ContentTypeNames.from( list );
    }

    public boolean canRender( ContentTypeName name )
    {
        return this.getCanRender().contains( name );
    }

    public boolean hasRegions()
    {
        return getPage().hasRegions();
    }

    public PageRegions getRegions()
    {
        return getPage().getRegions();
    }

    public static PageTemplate.Builder newPageTemplate()
    {
        return new Builder();
    }

    public static PageTemplate.Builder newPageTemplate( final PageTemplate source )
    {
        return new Builder( source );
    }

    public static PageTemplate.Builder copyOf( final PageTemplate pageTemplate )
    {
        return new Builder( pageTemplate );
    }

    public static class Builder
        extends Content.Builder<Builder, PageTemplate>
    {
        private PageTemplateKey key;

        private Builder()
        {
            super();
        }

        public Builder( final PageTemplate source )
        {
            super( source );
        }

        public Builder key( final PageTemplateKey value )
        {
            this.key = value;
            this.id = value;
            return this;
        }

        @Override
        public Content.Builder<Builder, PageTemplate> id( final ContentId value )
        {
            super.id( value );
            this.key = PageTemplateKey.from( value );
            return this;
        }

        public Builder controller( final DescriptorKey descriptorKey )
        {
            if ( this.page == null )
            {
                this.page = Page.newPage().
                    controller( descriptorKey ).
                    build();
            }
            else
            {
                this.page = Page.newPage( this.page ).
                    controller( descriptorKey ).
                    build();
            }
            return this;
        }

        public Builder canRender( final ContentTypeNames names )
        {
            for ( ContentTypeName name : names )
            {
                this.data.addProperty( "supports", Value.newString( name.toString() ) );
            }
            return this;
        }

        public Builder regions( final PageRegions value )
        {
            if ( this.page == null )
            {
                this.page = Page.newPage().
                    regions( value ).
                    build();
            }
            else
            {
                this.page = Page.newPage( this.page ).
                    regions( value ).
                    build();
            }

            return this;
        }

        public Builder config( final PropertyTree config )
        {
            if ( this.page == null )
            {
                this.page = Page.newPage().
                    config( config ).
                    build();
            }
            else
            {
                this.page = Page.newPage( this.page ).
                    config( config ).
                    build();
            }

            return this;
        }

        public PageTemplate build()
        {
            return new PageTemplate( this );
        }

    }

}

File: modules/admin-ui/src/main/resources/web/admin/common/js/content/page/PageTemplate.ts
module api.content.page {

    import ContentTypeName = api.schema.content.ContentTypeName;
    import Property = api.data.Property;
    import PropertyTree = api.data.PropertyTree;

    export class PageTemplate extends api.content.Content implements api.Equitable {

        private canRender: ContentTypeName[];

        constructor(builder: PageTemplateBuilder) {

            super(builder);

            this.canRender = [];
            this.getContentData().forEachProperty("supports", (property: Property) => {
                this.canRender.push(new ContentTypeName(property.getString()));
            });
        }

        getKey(): PageTemplateKey {

            return <PageTemplateKey>this.getContentId();
        }

        getPageMode(): api.content.page.PageMode {

            if (this.isPage()) {
                if (this.getPage().hasController()) {
                    return api.content.page.PageMode.FORCED_CONTROLLER;
                }
                else {
                    throw new Error("Illegal state: A PageTemplate's Page must a controller set");
                }
            }
            else {
                return api.content.page.PageMode.NO_CONTROLLER;
            }
        }

        getController(): DescriptorKey {

            return this.getPage().getController();
        }

        isCanRender(pattern: ContentTypeName): boolean {
            return this.getCanRender().some((name: ContentTypeName) => {
                return name.equals(pattern);
            });
        }

        getCanRender(): ContentTypeName[] {

            return this.canRender;
        }

        hasRegions(): boolean {
            if (!this.isPage()) {
                return false;
            }
            return this.getPage().hasRegions();
        }

        getRegions(): api.content.page.region.Regions {
            if (!this.isPage()) {
                return null;
            }
            return this.getPage().getRegions();
        }

        hasConfig(): boolean {
            if (!this.isPage()) {
                return false;
            }
            return this.getPage().hasConfig();
        }

        getConfig(): PropertyTree {
            if (!this.isPage()) {
                return null;
            }
            return this.getPage().getConfig();
        }

        equals(o: api.Equitable, ignoreEmptyValues: boolean = false): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, PageTemplate)) {
                return false;
            }

            return super.equals(o, ignoreEmptyValues);
        }

        clone(): PageTemplate {

            return this.newBuilder().build();
        }

        newBuilder(): PageTemplateBuilder {
            return new PageTemplateBuilder(this);
        }
    }

    export class PageTemplateBuilder extends api.content.ContentBuilder {

        constructor(source?: PageTemplate) {
            super(source);
        }

        fromContentJson(contentJson: api.content.json.ContentJson): PageTemplateBuilder {
            super.fromContentJson(contentJson);
            return this;
        }


        public build(): PageTemplate {
            return new PageTemplate(this);
        }
    }
}